java-client/src/main/java/co/elastic/clients/transport/ElasticsearchTransportBase.java

```diff
@@ -377,28 +379,42 @@ private <ResponseT> ResponseT decodeTransportResponse(
     ) throws IOException {
 
         if (endpoint instanceof JsonEndpoint) {
+
+            // Expecting a body
+            if (entity == null) {
+                throw new TransportException(
+                    clientResp,
+                    "Expecting a response body, but none was sent",
+                    endpoint.id()
+                );
+            }
+            InputStream content = entity.asInputStream();
+            InputStream contentForException = null;
+
+            // if the option to print the original body has been set, the body has to be
+            // copied first to another stream to be read again by the exception class
+            if(options().retrieveOriginalJsonResponseOnException()) {
+                try(ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+                    entity.writeTo(baos);
+                    content = new ByteArrayInputStream(baos.toByteArray());
+                    contentForException = new ByteArrayInputStream(baos.toByteArray());
+                }
+            }
+
             @SuppressWarnings("unchecked")
             JsonEndpoint<?, ResponseT, ?> jsonEndpoint = (JsonEndpoint<?, ResponseT, ?>) endpoint;
             // Successful response
             ResponseT response = null;
             JsonpDeserializer<ResponseT> responseParser = jsonEndpoint.responseDeserializer();
             if (responseParser != null) {
-                // Expecting a body
-                if (entity == null) {
-                    throw new TransportException(
-                        clientResp,
-                        "Expecting a response body, but none was sent",
-                        endpoint.id()
-                    );
-                }
```

Comment:
moved this up to check if the entity is null before trying to retrieve the content